salt/pillar/__init__.py

```diff
@@ -647,12 +656,12 @@ def get_tops(self):
             if self.opts["pillarenv"]:
                 # If the specified pillarenv is not present in the available
                 # pillar environments, do not cache the pillar top file.
-                if self.opts["pillarenv"] not in self.opts["pillar_roots"]:
+                if self.opts["pillarenv"] not in self._get_envs():
```

Comment:
Because when using `git_pillar`, the list of envs is not coming from `pillar_roots` but from `git`.

Note: As this is can be slow to list all the branches every time, I'm now storing this in `client._env`.